### PR TITLE
SCC-3601 - Advanced Search Utils

### PR DESCRIPTION
**Pull request by dgcohen**
This PR: 

Adds the qs (query string) library and uses it to decode the query string in the getServerSideProps function. With the addition of this library, it now supports decoding nested query string structures which allows the advanced search parameters to be parsed correctly.

Renames selectedFilters to filters (aesthetic choice, let me know if you disagree), and prepends a ? character in the return value of the query string builder (it was being prepended everywhere it was invoked)

Adds a POST request route handler in the Search Results API that maps the values from the request body to a query string (using a new helper utility) and redirects the user to the results page.

### CODE REVIEW

**Review by EdwinGuzman — COMMENTED** [review]
Somewhere, perhaps in the `getQueryString` function, the url should be encoded. I made the same advanced search query in the current RC and the Vercel instance and got two slightly different url query params:

```
// new
search?q=cat&contributor=mr+kittens&title=cats+are+nice&subject=pets&filters%5BdateBefore%5D=2022&filters%5BdateAfter%5D=2017&filters%5BmaterialType%5D%5B0%5D=resourcetypes%3Amix&filters%5BmaterialType%5D%5B1%5D=resourcetypes%3Anot

// current
search?q=cat&contributor=mr%20kittens&title=cats%20are%20nice&subject=pets&filters[dateAfter]=2017&filters[dateBefore]=2022&filters[materialType][0]=resourcetypes%3Amix&filters[materialType][1]=resourcetypes%3Anot
```

Looks like some strings are escaped but not all? How does qs handle the escaped brackets , for example in `materialType[1]` instead of `materialType%5D%5B1%5D`?

**Comment by EdwinGuzman** [comment]
Update: when I view the Vercel preview in an incognito window, the URL seems to be correct and matches what is on prod. It's a me problem.

File: src/utils/searchUtils.ts [hunk 1]
```diff
@@ -12,6 +12,8 @@ import type {
   Identifiers,
 } from "../types/searchTypes"
 
+// import { initialSearchFormState } from "./advancedSearchUtils"
```
Comment (EdwinGuzman) [inline src/utils/searchUtils.ts:15]: This can be removed.

**Review by EdwinGuzman — APPROVED** [review]
Some minor comments but looks good and works without js.

**Comment by charmingduchess** [comment]
> Looks like some strings are escaped but not all? How does qs handle the escaped brackets , for example in `materialType[1]` instead of `materialType%5D%5B1%5D`?

Was this question resolved?

**Comment by dgcohen** [comment]
> > Looks like some strings are escaped but not all? How does qs handle the escaped brackets , for example in `materialType[1]` instead of `materialType%5D%5B1%5D`?
> 
> Was this question resolved?

Yes, I wasn't able to replicate locally or on vercel and it seems to have been resolved on Edwin's end. We will keep an eye out on this in case we notice it.

File: pages/search/index.tsx [hunk 2]
```diff
@@ -74,8 +75,10 @@ export default function Search({ results }) {
   )
 }
 
-export async function getServerSideProps({ query }) {
-  const results = await fetchResults(mapQueryToSearchParams(query))
+export async function getServerSideProps({ resolvedUrl }) {
```
[the remark this answers is not in the document]
Comment (dgcohen) [reply pages/search/index.tsx:86]: Done!

**Pull request merged**